designs/component-upgrades.md

```diff
@@ -0,0 +1,149 @@
+
+We will need to add such field to all the other components (Flux, eks-a controllers and crds, cilium...) that don’t have it yet.
+
+### CAPI
+
+All CAPI providers can be updated with `clusterctl`. This works both with version bumps and new builds of the same version. We just need to create the proper overrides layer, configure the images in the clusterctl config file and pass specific versions for every provider to the `upgrade apply` command.
+
+Even if there is  CAPI api version change, the process is the same. The providers will take care of converting the objects in the cluster to the new api version. The only requirement is that all the installed providers need to be upgraded to a version that supports the new api version.
+
+### EKS-A controller and CRDs
+
+Both the controller and CRDs can be upgraded by just applying the new components manifest with kubectl. This should work fine for both new versions and new builds of the same version.
+
+We have to be careful and make sure that any changes to the CRDs are backwards compatible, since at this point there are already eks-a objects in the cluster. Examples of not backwards compatible changes: adding a required field, changing the type of field, making an existing field required, etc. It’s important to note that applying the manifest with the new CRD won’t fail, even if the changes are not backwards compatible. But the objects already in the cluster will become invalid, so any kind of update to those (eg. flux reconciliation) will be be rejected by the kube api server.
+
+Updates to the controller code should also take this into consideration. They always need to consider how to handle existing objects in the cluster. For example, if changing the pattern to name certain CAPI objects in the provider templates, the controller has to be able to handle both the old and the new pattern.
+
+If there is a new eks-a api version, we will need to write conversion webhooks.
+
+### Flux
+
+The process to upgrade the Flux components is very similar to the installation process, it uses the `flux` cli:
+
+1. Write the “kustomization” layer to disk to set our own Flux images
+2. Run the `bootstrap` command. This will generate the new manifest (`flux-system/gotk-components.yaml`) and commit all files to the repository
+3. Run the `reconcile` command for `flux-system` which tells Flux to pull the manifests from git and upgrade itself
+
+This process should work for both new versions and new builds of the same version.
+
+Full documentation: https://fluxcd.io/docs/installation/#bootstrap-upgrade
+
+### Cilium
+
+The process to upgrade cilium is a bit more involved: it uses helm to generate intermediate manifests and kubectl to apply them.
+
+1. Generate pre-flight check with helm and apply it with kubectl. This “pre-pulls" the new images on the nodes to avoid `ErrImagePull` errors later during the upgrade
```

Comment:
```suggestion
1. Generate pre-flight checks with `helm` and apply them with `kubectl`. This “pre-pulls" the new images on the nodes to avoid `ErrImagePull` errors later during the upgrade
```